### PR TITLE
chore(pkglint): disallow experimental dependencies

### PR DESCRIPTION
**Pull request by NetaNir**
> blocked on https://github.com/aws/aws-cdk/pull/14291

This PR adds a pkglint rule to disallow `@aws-cdk` modules to depend on experimental `@aws-cdk` modules. 

In v2 experimental modules will be released separately from `aws-cdk-lib`. This means that stable modules can not depend on experimental modules as it will create a cyclic dependency. The lint rule **also** ban exp->exp dependencies. The reasoning for the latter is that if an experimental module A, depends on experimental module B, in order to graduate module A, module B will also have to be graduated (to satisfy no stable->exp). Currently there are a few exp->exp dependencies, these might be removed as a part of the work to release the experimental modules separately from `aws-cdk-lib`. In the meantime we explicitly allow them by having an exclusion map.

The current implementation executes the check for each modules as a part of its `pkglint` phase, which is O(n^2) (for each module check the stability of all of its dependencies). This can be improved by executing the check only once, in the build of `aws-cdk-lib` (going over all the modules in a topological order and storing the stability). The disadvantage of executing it only on the build of `aws-cdk-lib` is that it will not be executed as a part of a typical developer flow, which is `buildup` from a specific package. I have testes it several times locally in the current implementation adds ~2 seconds to the root level `yarn pkglint`. Since it does not add significant time to our build, and gives better/earlier visibility to contributors I have decided to keep it simple and preform the check on each module build. Happy to discuss more if anyone have concerns or optimizations ideas I missed.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: tools/pkglint/lib/rules.ts [hunk 1]
```diff
@@ -1689,3 +1745,17 @@ function cdkMajorVersion() {
   const releaseJson = require(`${__dirname}/../../../release.json`);
   return releaseJson.majorVersion as number;
 }
+
+/**
+ * Should this package be included in the monolithic package.
+ */
+function isIncludedInMonolith(pkg: PackageJson): boolean {
```
Comment (NetaNir) [inline tools/pkglint/lib/rules.ts:1752]: @eladb I ended up duplicating this logic from `ubergen`. The logic in `ubergen` relies on the fact that the `ubergen.exclude` is not set on all modules that needs to excluded :/ . It's kinda of tricky, the **TL;DR** is that if the module is not `uber.exclude` there is a special logic that checks if it needs to be considered deprecated: [code](https://github.com/aws/aws-cdk/blob/master/tools/ubergen/bin/ubergen.ts#L104).

File: tools/pkglint/lib/rules.ts [hunk 2]
```diff
@@ -1616,13 +1616,68 @@ export class UbergenPackageVisibility extends ValidationRule {
...
+    }
+
+    Object.keys(pkg.dependencies).forEach(dep => {
+      if (!isCdkModuleName(dep)) {
+        return;
+      }
+
+      const shortName = dep.split('/')[1];
+      // eslint-disable-next-line @typescript-eslint/no-require-imports
+      const stability = require(path.resolve('../', shortName, 'package.json')).stability;
```
Comment (rix0rrr) [inline tools/pkglint/lib/rules.ts:1659]: Please don't write these kinds of rules that rely on physical locations of files in the source repository.

Instead, if you must find a dependency's `package.json`, do something like this:

```ts
const dependencyPj = path.resolve(`${shortName}/package.json`, { paths: [searchPath] });
const packageJson = await readJson(dependencyPj);
```

**Review by rix0rrr — CHANGES_REQUESTED** [review]

File: tools/pkglint/lib/rules.ts (hunk 2 above)
Comment (NetaNir) [reply tools/pkglint/lib/rules.ts]: So! We only need to read the `package.json` file of dependencies, which means they were already loaded during the build phase => we can require the dependency name, and we don't need a path:

```ts
const stability = require(`${dep}/package.json`).stability;
```

**Comment by NetaNir** [comment]
@rix0rrr thanks for the review, I have addressed your comment, can you have another look?

**Review by rix0rrr — APPROVED** [review]

**Pull request merged**